engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java

```diff
@@ -285,6 +288,7 @@ public IR.Expression apply(IR.Expression v1) {
             };
         var copy = this.ir.mapExpressions(fn);
         this.ir = copy;
+        this.uuidsMap = null;
```

Comment:
So does this mean that on every modification to the source we invalidate the whole cache? 
It's probably OK for now but I guess we could also in the future consider something more fine grained,

Reply:
Actually I was thinking about some support from @kazcw's parser - all the `UUIDs` are coming from the metadata section. The parser reads them all at once. However there is no access to all of them in the `Tree` API. That's why we currently have to `preorder()` thru all the `IR` nodes.

Having access to all of them as `Set<UUID>` would allow us to easily answer the question: is given `UUID` from this document or not?

> in the future consider something more fine grained,

In the future we would probably get incremental parser, which would also incrementally update the set of `UUID`s the document contains.